src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,94 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+
+    private List<Product> products;
+    private final Object lock = new Object();
```

Comment:
Please group fields by modifiers. It's common to declare fields from public to private, from static to non-static, from final to non-final.